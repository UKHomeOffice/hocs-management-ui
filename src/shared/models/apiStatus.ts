export default interface ApiStatus {
    display: string;
    timeoutPeriod: number;
    type: string;
}
