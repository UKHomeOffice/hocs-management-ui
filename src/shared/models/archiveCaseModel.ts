export default interface ArchiveCaseModel {
    caseReference: string;
    deleted: boolean;
}
