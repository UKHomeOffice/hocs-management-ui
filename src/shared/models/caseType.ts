export default interface CaseType {
    displayName: string;
    label: string;
    shortCode: string;
    type: string;
    value: string;
}
