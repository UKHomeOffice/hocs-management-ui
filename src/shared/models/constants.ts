export const ADD_CHILD_TOPIC_ERROR_DESCRIPTION = 'Something went wrong while adding the child topic. Please try again.';
export const ADD_STANDARD_LINE_ERROR_DESCRIPTION = 'Something went wrong while adding the standard line. Please try again.';
export const ADD_TEMPLATE_ERROR_DESCRIPTION = 'Something went wrong while adding the template. Please try again.';
export const ADD_UNIT_ERROR_DESCRIPTION = 'Something went wrong while adding the unit. Please try again.';
export const ADD_USER_ERROR_DESCRIPTION = 'Something went wrong while adding the following users. Please try again.';
export const ADD_USER_ERROR_TITLE = 'There was an error adding the users';
export const DUPLICATE_CHILD_TOPIC_DESCRIPTION = 'A child topic with that name already exists';
export const DUPLICATE_UNIT_DESCRIPTION = 'A unit with those details already exists';
export const EMPTY_SUBMIT_ERROR_DESCRIPTION = 'Please select some users before submitting.';
export const EMPTY_SUBMIT_ERROR_TITLE = 'No users selected';
export const EMPTY_SUBMIT_TEAM_ERROR_DESCRIPTION = 'Please select a team before submitting.';
export const EMPTY_SUBMIT_TEAM_ERROR_TITLE = 'No team selected';
export const EMPTY_SUBMIT_DRAFT_QA_TEAM_ERROR_DESCRIPTION = 'Select a team for Initial Draft and QA response stages.';
export const EMPTY_SUBMIT_PRIVATE_MINISTER_TEAM_ERROR_DESCRIPTION = 'Select a team for Private Office and Minister sign off stages.';
export const EMPTY_SUBMIT_TOPIC_ERROR_DESCRIPTION = 'Select a topic before submitting.';
export const EMPTY_SUBMIT_TOPIC_ERROR_TITLE = 'No topic selected';
export const GENERAL_ERROR_TITLE = 'Something went wrong';
export const LOAD_PARENT_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPIC_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_CASE_TYPES_DESCRIPTION = 'There was an error retrieving the case types.  Please try refreshing the page.';
export const LOAD_TEAM_ERROR_DESCRIPTION = 'There was an error retrieving the team.  Please try refreshing the page.';
export const LOAD_TEAM_MEMBERS_ERROR_DESCRIPTION = 'There was an error retrieving the teams members.  Please try refreshing the page.';
export const LOAD_TEAMS_ERROR_DESCRIPTION = 'There was an error retrieving the teams.  Please try refreshing the page.';
export const LOAD_USERS_ERROR_DESCRIPTION = 'There was an error retrieving the list of users.  Please try refreshing the page.';
export const REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION = 'The user cannot be removed from the team as they have cases assigned';
export const REMOVE_FROM_TEAM_ERROR_DESCRIPTION = 'There was an error removing the user from the team';
export const VALIDATION_ERROR_TITLE = "Thereâ€™s a problem";
