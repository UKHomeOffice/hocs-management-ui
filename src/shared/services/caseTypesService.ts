import axios from 'axios';
import CaseType from '../models/caseType';
import Item from '../models/item';

const sortByLabel: ((a: Item, b: Item) => number) = (a, b) => {
    const labelA = a.label.toLocaleLowerCase();
    const labelB = b.label.toLocaleLowerCase();
    let comparison = 0;
    if (labelA > labelB) {
        comparison = 1;
    } else if (labelA < labelB) {
        comparison = -1;
    }
    return comparison;
};

export const getCaseTypes = () => new Promise<CaseType[]>((resolve, reject) =>
    axios.get('/api/case-types')
        .then((response) => {
            resolve(response.data.sort(sortByLabel));
        })
        .catch(error => reject(error))
);
