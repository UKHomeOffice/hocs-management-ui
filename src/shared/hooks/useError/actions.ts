import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
