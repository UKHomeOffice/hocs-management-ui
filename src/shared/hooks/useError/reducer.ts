import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
