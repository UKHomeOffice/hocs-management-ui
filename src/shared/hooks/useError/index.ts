import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
