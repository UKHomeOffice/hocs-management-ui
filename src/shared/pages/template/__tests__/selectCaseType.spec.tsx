import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, render, RenderResult, wait, fireEvent, waitForElement } from '@testing-library/react';
import SelectCaseType from '../selectCaseType';
import { GENERAL_ERROR_TITLE, LOAD_CASE_TYPES_DESCRIPTION } from '../../../models/constants';
import Item from '../../../models/item';
import * as useError from '../../../hooks/useError';
import * as CaseTypesService from '../../../services/caseTypesService';

jest.mock('../../../services/caseTypesService', () => ({
    __esModule: true,
    getCaseTypes: () => Promise.resolve([{
        label: '__caseType1__',
        value: '__CaseTypeId1__'
    }, {
        label: '__caseType2__',
        value: '__CaseTypeId2__'
    }])
}));

const getCaseTypesSpy = jest.spyOn(CaseTypesService, 'getCaseTypes');

let match: match<any>;
let history: History<any>;
let location: Location;
let wrapper: RenderResult;
let mockSelectedCaseType: Item | undefined;

const useStateSpy = jest.spyOn(React, 'useState');
const setSelectedCaseTypeMock = jest.fn();
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorMock = jest.fn();
const clearErrorsMock = jest.fn();
const setMessageMock = jest.fn();
mockSelectedCaseType = undefined;

const renderComponent = () => render(
    <MemoryRouter>
        <SelectCaseType history={history} location={location} match={match}></SelectCaseType>
    </MemoryRouter>
);

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: { teamId: '__teamId__' },
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };

    useStateSpy.mockImplementation(() => [mockSelectedCaseType, setSelectedCaseTypeMock]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorMock, clearErrorsMock, setMessageMock]);
    history.push = jest.fn();
    addFormErrorMock.mockReset();
    clearErrorsMock.mockReset();
    setMessageMock.mockReset();
    act(() => {
        wrapper = renderComponent();
    });
});

describe('when the selectCaseType component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(1);

        await wait(() => {
            expect(wrapper.container).toMatchSnapshot();
        });
    });

    it('should display an error if the call to retrieve the case types fail', async () => {
        expect.assertions(1);
        getCaseTypesSpy.mockImplementation(() => Promise.reject('error'));

        act(() => {
            wrapper = renderComponent();
        });

        await wait(() => {
            expect(setMessageMock).toBeCalledWith({ title: GENERAL_ERROR_TITLE, description: LOAD_CASE_TYPES_DESCRIPTION });
        });

    });
});

describe('when the submit button is clicked', () => {
    describe('and the data is filled in', () => {

        beforeAll(async () => {
            mockSelectedCaseType = {
                label: '__caseTypeLabel__',
                value: '__caseTypeValue__'
            };
        });

        beforeEach(async () => {
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should redirect to the home page', async () => {
            expect.assertions(1);

            await wait(() => {
                expect(history.push).toHaveBeenCalledWith('/case-type/__caseTypeValue__');
            });
        });
        it('should clear any previous errors', async () => {
            expect.assertions(1);

            await wait(() => {
                expect(clearErrorsMock).toHaveBeenCalled();
            });
        });
    });
    describe('and the data is not filled in', () => {

        beforeAll(async () => {
            mockSelectedCaseType = undefined;
        });
        beforeEach(async () => {
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect(clearErrorsMock).toHaveBeenCalled();
        });

        it('should set the error state', () => {
            expect(addFormErrorMock).toHaveBeenNthCalledWith(1, { key: 'value', value: 'The Case Type is required' });
        });
    });
});
