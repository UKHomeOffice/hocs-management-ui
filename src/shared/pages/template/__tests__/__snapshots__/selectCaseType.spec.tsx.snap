// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the selectCaseType component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Select a Case Type
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        action="/api/template"
        method="POST"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="case-types-input"
            id="case-types-label"
          >
            Case Types
          </label>
          <div
            class=" css-2b097c-container"
            id="case-types"
          >
            <div
              class="govuk-typeahead__control css-1g8ily9-Control"
            >
              <div
                class="govuk-typeahead__value-container css-pilx6v-ValueContainer"
              >
                <div
                  class="govuk-typeahead__placeholder css-ovmm9z-Placeholder"
                >
                  Search
                </div>
                <div
                  class="css-b8ldur-Input"
                >
                  <div
                    class="govuk-typeahead__input"
                    style="display: inline-block;"
                  >
                    <input
                      aria-autocomplete="list"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      id="case-types-input"
                      spellcheck="false"
                      style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                    <div
                      style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                    />
                  </div>
                </div>
              </div>
              <div
                class="govuk-typeahead__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="govuk-typeahead__indicator-separator css-62wm9a-IndicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="govuk-typeahead__indicator govuk-typeahead__dropdown-indicator css-hbt8bd-DropdownIndicator"
                >
                  <svg
                    aria-hidden="true"
                    class="css-6q0nyr-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;
